lib/Storage/MongoDbTimeSeriesStore.php

```diff
@@ -120,137 +127,231 @@ public function setDependencies($log, $date, $layoutFactory = null, $campaignFac
-                    if (isset($arrayOfTags[$i]) && (isset($arrayOfTagValues[$i]) && $arrayOfTagValues[$i] !== 'NULL' )) {
-                        $tagFilter['dg'][$i]['tag'] = $arrayOfTags[$i];
-                        $tagFilter['dg'][$i]['val'] = $arrayOfTagValues[$i];
-                    } else {
-                        $tagFilter['dg'][$i]['tag'] = $arrayOfTags[$i];
-                    }
+        $statData['layoutName'] = $layoutName;
+
+        // Layout tags
+        if (isset($layoutTags)) {
+            $arrayOfTags = explode(',', $layoutTags);
+            for ($i=0; $i<count($arrayOfTags); $i++) {
+                if (isset($arrayOfTags[$i]) ) {
+
+                    if (!empty($arrayOfTags[$i]))
+                        $tagFilter['layout'][$i]['tag'] = $arrayOfTags[$i];
                 }
-            } catch (NotFoundException $notFoundException) {
-                $this->log->alert('Error processing statistic into MongoDB. Display not found. Stat is: ' . json_encode($stat));
-                // TODO: need to remove the record?
-                continue;
             }
-
-            // TagFilter array
-            $statData[$k]['tagFilter'] = $tagFilter;
         }
 
+        // Display tags
+       if (array_key_exists($display->displayGroupId, $this->tagFilterDg)) {
+
+           if (isset($this->tagFilterDg[$display->displayGroupId])) {
+               $tagFilter['dg'] = $this->tagFilterDg[$display->displayGroupId];
+           }
+
+       } else {
+
+           try {
+
+               $arrayOfTags = array_filter(explode(',', $this->displayGroupFactory->getById($display->displayGroupId)->tags));
+               $arrayOfTagValues = array_filter(explode(',', $this->displayGroupFactory->getById($display->displayGroupId)->tagValues));
+
+               for ($i=0; $i<count($arrayOfTags); $i++) {
+                   if (isset($arrayOfTags[$i]) && (isset($arrayOfTagValues[$i]) && $arrayOfTagValues[$i] !== 'NULL' )) {
+                       $tagFilter['dg'][$i]['tag'] = $arrayOfTags[$i];
+                       $tagFilter['dg'][$i]['val'] = $arrayOfTagValues[$i];
+                   } else {
+                       $tagFilter['dg'][$i]['tag'] = $arrayOfTags[$i];
+                   }
+               }
+
+               // Put in memory
+               $this->tagFilterDg[$display->displayGroupId] = $tagFilter['dg'];
+
+
+           } catch (NotFoundException $notFoundException) {
+               $this->log->alert('Error processing statistic into MongoDB. Display not found. Stat is: ' . json_encode($statData));
+               return;
+           }
+       }
+
+        // TagFilter array
+        $statData['tagFilter'] = $tagFilter;
+
+        $this->stats[] = $statData;
+
+    }
+
+    /** @inheritdoc */
+    public function addStatFinalize()
+    {
+
         // Insert statistics
         $collection = $this->client->selectCollection($this->config['database'], $this->table);
```

Comment:
I didn't notice this before, but I think we should throw an error out after we log here?

If the SubmitStats request returns a 200 response, the Player will delete the stats locally - but if the insert fails here, we should make the player try again I think? we do that be throwing an exception

I think the MySQL one does throw an exception, but worth checking.

Reply:
$this->store->isolated(..)  does throw PDOException

Reply:
I think it does actually - if its not a reconnect error it catches and rethrows